### ISSUE TITLE
Build for 2.13.0

### ISSUE THREAD

**Issue by rnd4222** [twitter/finagle#771]
Scala 2.13.0 was [released](https://github.com/scala/scala/releases/tag/v2.13.0), so it would be great to have finagle build for it.

If it's marked with an asterisk, it doesn't have any dependencies and will be easy to target next.  If you want to discover what the dependencies of a given project are, take a peek inside of the build.sbt file.

If you start working on one of these subprojects, please add a comment on this thread and I'll add your name next to it so people know not to duplicate your work.

- [x] finagle-base-http* (@martijnhoekstra)
- [x] finagle-benchmark-thrift* (@joroKr21)
- [x] finagle-benchmark(@martijnhoekstra)
- [x] finagle-core (@martijnhoekstra)
- [x] finagle-exception (@DieBauer)
- [x] finagle-exp (@DieBauer)
- [x] finagle-grpc-context* (@martijnhoekstra)
- [x] finagle-http (@martijnhoekstra, @spockz)
- [x] finagle-http2 (@martijnhoekstra, @spockz)
- [x] finagle-init  (@martijnhoekstra)
- [x] finagle-integration (@felixbr)
- [x] finagle-memcached(@martijnhoekstra)
- [x] finagle-mux* (@martijnhoekstra)
- [x] finagle-mysql (@martijnhoekstra)
- [x] finagle-netty4 (@martijnhoekstra)
- [x] finagle-netty4-http (@martijnhoekstra, @spockz)
- [x] finagle-opencensus-tracing(@hderms)
- [x] finagle-partitioning (@martijnhoekstra)
- [x] finagle-redis (@martijnhoekstra)
- [x] finagle-serversets* (@joroKr21)
- [x] finagle-stats (@hderms)
- [x] finagle-stats-core (@hderms)
- [x] finagle-thrift (@martijnhoekstra)
- [x] finagle-thriftmux (@martijnhoekstra)
- [x] finagle-toggle  (@martijnhoekstra)
- [x] finagle-tunable (@martijnhoekstra)
- [x] finagle-zipkin-core (@DieBauer)
- [x] finagle-zipkin-scribe (@DieBauer)

**Comment by erwan** [comment]
There is at least a dependency on utils: https://github.com/twitter/util/pull/248

Scalatest also needs to be updated, probably other dependencies as well

**Comment by yaroot** [comment]
and twitter/scrooge#308

**Comment by luciferous** [comment]
As this will be a lot of work, we plan to coordinate it with the community in a similar fashion to how we did 2.11 (#290) and 2.12 (#505).

**Comment by SethTisue** [comment]
the next step here is scrooge, I believe

**Comment by mosesn** [comment]
we can start work on some of the pieces, like finagle-core (basically anything that doesn't depend on scrooge within finagle), if people want to start taking stabs at subprojects!

**Comment by mosesn** [comment]
Now that everything in scrooge is working except for scrooge-generator-tests, we're unblocked on everything in finagle.

**Comment by clhodapp** [comment]
What is the significance of the asterisks on items in the issue description?

**Comment by martijnhoekstra** [comment]
If I had to guess, I'd say those are not blocked by dependencies and open to port atm.

**Comment by mosesn** [comment]
Yep, @martijnhoekstra is right. I updated the original post to note that, sorry for the confusion

**Comment by mosesn** [comment]
If you start working on one of these subprojects, please add a comment on this thread and I'll add your name next to it so people know not to duplicate your work.

**Comment by martijnhoekstra** [comment]
thrift, mux and thriftmux at https://github.com/twitter/finagle/pull/794

**Comment by martijnhoekstra** [comment]
following up on base-http - the biggest issue I'm running in to is Header. This is conceptually a mutable multimap case-insentitive string -> string. It accomplishes this by overriding the elementHash and elementEquality. This is no longer possible in 2.13. Hopefully someone sees an option beyond using a different (hand-rolled? copy-pasted? 3rd party dependency?) map implementation, or wrapping all keys and overriding equals and hashcode, because those options are pretty unappealing.

**Comment by mosesn** [comment]
Sorry, I just saw this now.  Let me talk to the team and see if I can get a consensus on what other testing we would have to do to be comfortable with that kind of change.

**Comment by mosesn** [comment]
@martijnhoekstra are you imagining that we would stop extending HashMap?

**Comment by martijnhoekstra** [comment]
@mosesn yes, my current plan is to re-implement `Headers` which is used as the backing store for `DefaultHeaderMap` to become a mutable prefix tree on case-insentive char values, with `Header` at the leaves, and then run the benchmarks to check my suspicion that it's a strict win in terms of performance.

`Headers` is not exposed publicly, or even internally really, and this change shouldn't (TM) change any visible interface.

**Comment by martijnhoekstra** [comment]
@mosesn there is a very WIP-y WIP in https://github.com/martijnhoekstra/finagle/tree/http_base_213

The mutable Headers trie implementation is https://github.com/martijnhoekstra/finagle/blob/http_base_213/finagle-base-http/src/main/scala/com/twitter/finagle/http/Headers.scala

I'm currently working on getting 2.12 correct on this design so that I can bench against the baseline.

**Comment by mosesn** [comment]
The trie approach seems pretty simple so far.  Another idea folks were floating internally was to just go with a List and call it a day.  Did you try benchmarking that?

**Comment by martijnhoekstra** [comment]
just a List[Header], with linear search add/put/remove and toLowerCase'ing both sides? The original code did try very hard to be clever enough to be fast, but I can give it a trie. I haven't benchmarked a thing yet.

**Comment by mosesn** [comment]
well since you're building the trie anyway, let's start with that and see how it goes, and see how much worse List[Header] is?

**Comment by martijnhoekstra** [comment]
The current trie inplementation I have is hilariously slow on iteration - 10x vs the baseline. I probably messed something up.

**Comment by martijnhoekstra** [comment]
Looking for guiding review on https://github.com/twitter/finagle/pull/797

**Comment by DieBauer** [comment]
I've addressed `finagle-exp` in https://github.com/twitter/finagle/pull/802

**Comment by DieBauer** [comment]
And took on `finagle-zipkin-core` in #803

**Comment by DieBauer** [comment]
`finagle-zipkin-scribe` is also added to #803, to kind of belong together :)

**Comment by joroKr21** [comment]
PR for `finagle-serversets`: #810

**Comment by joroKr21** [comment]
`finagle-benchmark-thrift`: #811

**Comment by spockz** [comment]
* finagle-netty4-http
* finagle-http2
* finagle-http

Available in:

https://github.com/martijnhoekstra/finagle/pull/1

**Comment by chrisbenincasa** [comment]
Any update on when this will be available?

**Comment by yufangong** [comment]
Hi @chrisbenincasa, the progress of this is up-to-date, we just merged in **finagle-base-http**. We are taking an OSS community collaboration approach for this work so we don't have an accurate estimation, sorry.

**Comment by chrisbenincasa** [comment]
Thanks for the update, Yufan. Is there anything I can do to help?

**Comment by ryanoneill** [comment]
@chrisbenincasa if you want to take one of the unaccounted for libraries listed above, that would be great.

**Comment by yufangong** [comment]
to add on what @ryanoneill said, the one with * (right now finagle-memcached*) means it should not be blocked by other dependencies.

**Comment by chrisbenincasa** [comment]
Gotcha, thanks. Was looking into memcached a bit today but it looks like it's blocked on 2.13 in bijection (handled here https://github.com/twitter/bijection/pull/287 by @martijnhoekstra). Other than that, it seems other modules are still blocked?

**Comment by hderms** [comment]
@martijnhoekstra I am going to try to tackle `finagle-stats` and `finagle-stats-core`

**Comment by yufangong** [comment]
@chrisbenincasa Thanks for looking at it, good find, we will need to work on the bijection dep first. I took another look at the project list, I think the `finagle-opencensus-tracing` might be unblocked by the recent http projects.

**Comment by hderms** [comment]
@martijnhoekstra @yufangong I am handling finagle-opencensus-tracing as well https://github.com/twitter/finagle/pull/822

**Comment by hderms** [comment]
How can we identify whether there will be performance regressions on a given upgrade like what happened in https://github.com/twitter/finagle/pull/797 ? Is there a benchmark that is run automatically and set to fail if performance falls below some threshold, or some other mechanism I can use to determine if one of my update PRs regresses? 

Curious because I'm about to do `finagle-memcached`, and did `finagle-stats` before, and both seem like the kind of modules that could actually experience some kind of measurable performance regression.

**Comment by martijnhoekstra** [comment]
twitter-bijection 2.13 support was merged, but it's not part of the dodo build, so when a release comes around, maybe the last loose ends can be tied down.

**Comment by chrisbenincasa** [comment]
It seems we are waiting on the version of bijection to be updated in the private version of this repo before we can move forward, right? Is that the last blocker here?

**Comment by martijnhoekstra** [comment]
No, bijection is through. There is a PR for memcached and benchmarks:  https://github.com/twitter/finagle/pull/827

That blocks integration and example. I think that's the last of it.

Then I think it would be nice to get all CI on 2.13 too.

**Comment by chrisbenincasa** [comment]
Gotcha, sorry, I fell a bit behind on what was happening here. Is there anything I can do to help?

**Comment by martijnhoekstra** [comment]
I have no idea how finagle-doc works, but it's not blocked by any subprojects anymore. Maybe that can be upgraded too.

Cross-project, twitter/twitter-server doesn't seem blocked anymore either. I'm not entirely sure what's still left to do for other projects, maybe it would be good to make a matrix for what's left to do and what's blocked (by what) for all OSS projects. For example, Finatra is blocked by twitter-server, and there is no CI for scrooge on 2.13. Having an overview of all project statuses would definitely help to find the places that are unblocked for further progress, and what is still todo but blocked.

**Comment by felixbr** [comment]
> twitter/twitter-server doesn't seem blocked anymore either

Technically it isn't blocked anymore as all finagle modules it uses are available for 2.13. Because of the way `dodo` works, however, it needs the full finagle project to build for 2.13, so the CI currently fails.

PR for reference: https://github.com/twitter/twitter-server/pull/60

**Comment by felixbr** [comment]
Just to keep track of everything:
`finagle-benchmark` is still marked as missing but it seems it was already included in #827.

This leaves only `finagle-integration`. 
I can take a stab at it if no one else has already started on it :)

**Comment by yufangong** [comment]
@felixbr thanks! it seems covered in finagle-memcached. Please go ahead with finagle-integration. I'm planning to do a March release soon and willing to wait for all subjects on 2.13.

**Comment by martijnhoekstra** [comment]
There also is the otherwise unmentioned finagle-doc which looks to depend on other tooling (I didn't try getting it to build or run for very long)

**Comment by martinffx** [comment]
@yufangong did this make it into the release?

**Comment by yufangong** [comment]
@martinffx this will be in April release which will happen in the next couple days.

**Comment by yufangong** [comment]
We believe the most recent release 20.4.1 should have everything cross-building with Scala 2.13, I'm going to close this issue.
Thanks all for making this happen!!! ❤️ ❤️ ❤️
